### PR TITLE
22: Multiplayer

### PR DESCRIPTION
**Pull request by Wilson-Orlando**
Implemented multiplayer for up to three players, as well as Opponent's Choice sector.

Branched from Robin's #25 branch; we should merge that one first.

### CODE REVIEW

File: client.py [hunk 1]
```diff
@@ -119,7 +119,7 @@ def handle_connection(self, server):
                 response_info = [chosen_category]
 
             elif parsed_message.code == MessageType.SPIN:
-                [player_id] = parsed_message.args  # TODO: Why is this client receiving its own player ID?
+                _ = parsed_message.args
```
Comment (robyntorregrosa) [inline client.py:122]: I think we should have the server send the `player_id` of the player taking the turn, that way the client can check if that is their ID and decide whether to display the "spin" button/prompt.

File: executive_logic/executive_logic.py [hunk 2]
```diff
@@ -206,40 +213,35 @@ def store_query(self, command: MessageType, args: list):
...
 
-    def __update_scores_tokens_spins(self):
+    def __update_scores_tokens_spins(self, curr_player_id):
         """
         Called after a game value is updated that should be reflected in the UI display.
         Sends server the current player scores, player tokens, and number of spins remaining.
         :return: void
         """
-        self.__query_server(MessageType.UPDATE_SCORES,
+        self.__query_server(curr_player_id, MessageType.UPDATE_SCORES,
```
Comment (robyntorregrosa) [inline executive_logic/executive_logic.py:222]: Are you only notifying the current player of the updated score? I think the scoreboard should be updated for all the players still.

Comment (robyntorregrosa) [reply executive_logic/executive_logic.py:222]: I'm seeing now that update_scores calls __query_server which waits for a response from the client. It probably makes sense for the "update" type messages to call a different function that just pushes the updated and doesn't make the server wait for a response from the client.s

File: executive_logic/executive_logic.py [hunk 3]
```diff
@@ -206,40 +213,35 @@ def store_query(self, command: MessageType, args: list):
...
-    ########
-    # TODO: All methods below this point may be deprecated?
-    ########
-    def notify_spin(self):
-        self.num_spins += 1
-        return self.num_spins
-
-    def whose_turn(self,
-                   curr_player_id):  # TODO: This method may require changing curr_player_id to a class variable, or passing the information down from Executive Logic to the class that needs it. Revisit later.
+    def __notify_all_players(self, command: MessageType, args: list):
```
Comment (robyntorregrosa) [inline executive_logic/executive_logic.py:226]: Write `__notify_all_players` so it doesn't require a response from the clients.

File: client.py [hunk 4]
```diff
@@ -61,7 +61,7 @@ def handle_connection(self, server):
                 raise Exception(f"This client (ID %s) was already assigned a player ID!", self.player_id)
 
             elif parsed_message.code == MessageType.JEOPARDY_QUESTION:
-                [player_id, jeopardy_category, tile] = parsed_message.args  # TODO: Why is this client receiving its own player ID?
+                [jeopardy_category, tile] = parsed_message.args
```
Comment (robyntorregrosa) [inline client.py:64]: I would leave `playerID` here so the client can figure out if it is currently their turn and then only display the "submit" button if it is their turn.

File: executive_logic/executive_logic.py (hunk 3 above)
Comment (Wilson-Orlando) [reply executive_logic/executive_logic.py:226]: Reposting my response on discord: I don't think there's an easy way to do that. Messenger::buffer_message is the only way for the server to interface with the client, and it requires a response from the client.

The workaround is for the client to return an empty list, like with SPIN_RESULT. For __notify_all_players, it's just sending messages like that in series.

**Pull request merged**